integration/querier_test.go

```diff
@@ -471,38 +470,43 @@ func TestQuerierWithBlocksStorageRunningInSingleBinaryMode(t *testing.T) {
 			require.NoError(t, cluster.WaitSumMetrics(e2e.GreaterOrEqual(float64(shippedBlocks*seriesReplicationFactor)), "cortex_bucket_store_blocks_loaded"))
 
 			var expectedCacheRequests int
+			var expectedCacheHits int
+			var expectedMemcachedOps int
 
 			// Query back the series (1 only in the storage, 1 only in the ingesters, 1 on both).
 			// For every series in a block we expect
 			// - 1 cache request for expanded postings
 			// - 1 cache request for postings for each matcher (provided matchers are simple MatchEQ matchers)
-			// - 1 cache request for each series
+			// - 2 cache requests for each series: one while sending label values and another while streaming chunks. The second request should be a cache hit.
 			//
 			// If the series does not exist in the block, then we return early after checking expanded postings and
 			// subsequently the index on disk.
 			result, err := c.Query(series1Name, series1Timestamp)
 			require.NoError(t, err)
 			require.Equal(t, model.ValVector, result.Type())
 			assert.Equal(t, expectedVector1, result.(model.Vector))
-			expectedCacheRequests += seriesReplicationFactor * 3 // expanded postings, postings, series
+			expectedCacheRequests += seriesReplicationFactor * 4        // expanded postings, postings, series (retrieving labels and then while streaming chunks)
+			expectedCacheHits += seriesReplicationFactor                // one for each series (for chunks)
+			expectedMemcachedOps += seriesReplicationFactor * (3*2 + 1) // Same reasoning as for expectedCacheRequests, but this also includes a set for each get that is not a hit
```

Comment:
isn't this effectively the sum of the increases in expectedCacheRequests and expectedCacheHits?

```suggestion
			expectedMemcachedOps += seriesReplicationFactor * (3*2 + 1) // Same reasoning as for expectedCacheRequests and expectedCacheHits, but this also includes a set for each get that is not a hit
```

Reply:
Turns out it's a bit more complicated than that: the number of Memcached ops is not predictable, because there's a race between the two cache operations while streaming series - set operations are async, and so the first set operation triggered while sending labels may not have completed by the time streaming chunks tries to do a get on the same item.

This also means that the number of hits is not predictable either.

I'm now wondering if we're doubling up work if this happens in the real world, and if there's anything we can do about it...

Reply:
This issue was fixed in https://github.com/grafana/mimir/pull/8039.